Add basic token interface

shadowhand opened the pull request:
https://twitter.com/frankdejonge/status/634727793805406209
https://groups.google.com/forum/#!msg/thephpleague/yZkYhIr3pS4/DqolYc3kGQAJ

cc @frankdejonge @alexbilbie @ramsey @stevenmaguire @sammyk

shadowhand committed:
Add basic token interface

alexbilbie commented:
Have a look at how I've implemented tokens in oauth2-server v5 https://github.com/thephpleague/oauth2-server/tree/V5-WIP/src/Entities

shadowhand commented:
@alexbilbie as per the discussions we had on the ML, the first step is a baseline token interface that can be shared with v1 and v2 tokens. If all interested parties are okay with what I have here, I will start adding more interfaces.

At a quick glance, my primary concern with the interfaces/implementations that you've created for oauth2-server is that they are not immutable.

stevenmaguire commented:
I like it. Seems pretty straight forward. I like the pattern consistency from the PSR 7 interfaces which create copies of objects as their values change. :+1: from me.

alexbilbie commented:
I think the methods should be called `getIdentifier` and `withIdentifier`; the token represents a payload shared between the client and server. Some implementations use the token itself as the identifier that is looked up to verify it's validity; others (such as JWTs) will contain an embedded identifier that is extracted for lookups.

ramsey commented:
I think we should stick with the nomenclature used by OAuth, which is "token." The word "identifier" carries with it other meaning, and it may not be used as a unique identifier at all.

shadowhand committed:
Add OAuth2 access token implementation

shadowhand commented:
I've updated this PR with an implementation of OAuth2 access tokens and added tests for the classes. Have I made any omissions? Will this style work correctly with multiple implementations?

I would also like to add `public function toArray() : array` that returns a structure that matches the spec and possibly also a `public static function fromArray(string $token, string $scopeSeparator = ' ') : AccessToken` that would make it easier to create the token from a raw response.
